src/Lib.P2P/AutoDialer.cs

```diff
@@ -0,0 +1,159 @@
+        /// </summary>
+        /// <param name="swarm">
+        ///   Provides access to other peers.
+        /// </param>
+        public AutoDialer(Swarm swarm)
+        {
+            this.swarm = swarm;
+            swarm.PeerDiscovered += OnPeerDiscovered;
+            swarm.PeerDisconnected += OnPeerDisconnected;
+        }
+
+        /// <summary>
+        ///  Releases the unmanaged and optionally managed resources.
+        /// </summary>
+        /// <param name="disposing">
+        ///   <b>true</b> to release both managed and unmanaged resources; <b>false</b> 
+        ///   to release only unmanaged resources.
+        /// </param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                swarm.PeerDiscovered -= OnPeerDiscovered;
+                swarm.PeerDisconnected -= OnPeerDisconnected;
+            }
+        }
+
+        /// <summary>
+        ///   Performs application-defined tasks associated with freeing, 
+        ///   releasing, or resetting unmanaged resources.
+        /// </summary>
+        public void Dispose() { Dispose(true); }
+
+        /// <summary>
+        ///   The low water mark for peer connections.
+        /// </summary>
+        /// <value>
+        ///   Defaults to <see cref="DefaultMinConnections"/>.
+        /// </value>
+        /// <remarks>
+        ///   Setting this to zero will basically disable the auto dial features.
+        /// </remarks>
+        public int MinConnections { get; set; } = DefaultMinConnections;
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        /// <summary>
+        ///   Called when the swarm has a new peer.
+        /// </summary>
+        /// <param name="sender">
+        ///   The swarm of peers.
+        /// </param>
+        /// <param name="peer">
+        ///   The peer that was discovered.
+        /// </param>
+        /// <remarks>
+        ///   If the <see cref="MinConnections"/> is not reached, then the
+        ///   <paramref name="peer"/> is dialed.
+        /// </remarks>
+        private async void OnPeerDiscovered(object sender, Peer peer)
+#pragma warning restore VSTHRD100 // Avoid async void methods
+        {
+            var n = swarm.Manager.Connections.Count() + pendingConnects;
+            if (swarm.IsRunning && n < MinConnections)
+            {
+                Interlocked.Increment(ref pendingConnects);
+                log.Debug($"Dialing new {peer}");
+                try
+                {
+                    await swarm.ConnectAsync(peer).ConfigureAwait(false);
+                }
+                catch (Exception)
+                {
+                    log.Warn($"Failed to dial {peer}");
+                }
+                finally
+                {
+                    Interlocked.Decrement(ref pendingConnects);
+                }
+            }
+        }
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not suppress issues.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)